worldbank/DECAT_Space2Stats
# Bugfix/58/h3 parent approach

## What I Changed

- Implemented the `_find_touches` function to identify H3 hexagon IDs that intersect with the boundary of the polygon defined by the AOI GeoJSON and include neighboring hexagons.
- Refactored the `_recursive_polyfill` method to handle cases where no H3 IDs are returned, allowing it to continue to higher resolutions until a valid result is found.
- Enhanced the logic to ensure the correct filtering of H3 hexagon IDs based on spatial join methods: `within` and `touches`.

## How to Test It

- See `test_h3_utils.py`
- Use the `generate_h3_ids` function with various Polygon GeoJSON inputs to ensure that H3 IDs are generated correctly for different geographic shapes.
- Verify that the filtering based on spatial join methods works as expected, particularly with the new `_find_touches` function.
- Run the existing unit tests to confirm that no regressions have occurred in the functionality.

Quick example running with the `touches` keyword:
![image](https://github.com/user-attachments/assets/93811384-de16-4b3a-b8a3-4510b6e6eecd)

## Other Notes

- Closes #58
- I somehow started from a branch that didn't have the support for multipolygon integrated. I'll resolve the merge conflict and see how to add multipolygon support tomorrow. cc: @alukach
